templates/commands/render.go

```diff
@@ -235,24 +275,237 @@ func (r *Render) realRun(ctx context.Context, rp *runParams) (outErr error) {
+			if errors.Is(err, os.ErrNotExist) {
+				return errWithPos(p.Pos, "include path doesn't exist: %q", walkRelPath)
+			}
+		}
+
+		if err := copyRecursive(p.Pos, absSrc, absDst, sp.fs); err != nil {
+			return errWithPos(p.Pos, "copying failed: %w", err)
+		}
+	}
+	return nil
+}
+
+// "from" may be a file or directory. "pos" is only used for error messages.
+func copyRecursive(pos *model.ConfigPos, from, to string, rfs renderFS) error {
+	return fs.WalkDir(rfs, from, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err // There was some filesystem error. Give up.
+		}
+
+		// We don't have to worry about symlinks here because we passed
+		// DisableSymlinks=true to go-getter.
+
+		relToSrc, err := filepath.Rel(from, path)
+		if err != nil {
+			return errWithPos(pos, "filepath.Rel(%s,%s): %w", from, path, err)
+		}
+
+		dst := filepath.Join(to, relToSrc)
+
+		// The spec file may specify a file to copy that's deep in a
+		// directory tree, without naming its parent directory. We can't
+		// rely on WalkDir having traversed the parent directory of $path,
+		// so we must create the target directory if it doesn't exist.
+		dirToCreate := dst
+		if !d.IsDir() {
+			dirToCreate = filepath.Dir(dst)
+		}
+		if err := rfs.MkdirAll(dirToCreate, 0o700); err != nil {
+			return errWithPos(pos, "MkdirAll(): %w", err)
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		buf, err := rfs.ReadFile(path)
+		if err != nil {
+			return errWithPos(pos, "ReadFile(): %w", err)
+		}
+
+		// The permission bits for normal files in the scratch directory will be
+		// set conservatively to either 600 (u=rw) or 700 (u=rwx), depending on
+		// the owner-executable permission bit of the source file. These
+		// permissions will be further limited by the user's umask (see `man
+		// umask` if unfamiliar), so the actual permissions on disk might be
+		// more restrictive than we ask for.
+		info, err := rfs.Stat(path)
+		if err != nil {
+			return fmt.Errorf("Stat(): %w", err)
+		}
+		ownerExecuteBit := info.Mode() & 0o100
+		dstPerms := 0o600 | ownerExecuteBit
+
+		if err := rfs.WriteFile(dst, buf, dstPerms); err != nil {
+			return fmt.Errorf("failed writing to scratch file: WriteFile(): %w", err)
+		}
+
+		return nil
+	})
+}
+
+func actionRegexReplace(ctx context.Context, p *model.RegexReplace, sp *stepParams) error {
+	return fmt.Errorf("not implemented")
+}
+
+func actionStringReplace(ctx context.Context, p *model.StringReplace, sp *stepParams) error {
+	return fmt.Errorf("not implemented")
+}
+
+func actionGoTemplate(ctx context.Context, p *model.GoTemplate, sp *stepParams) error {
+	return fmt.Errorf("not implemented")
+}
+
+func loadSpecFile(fs renderFS, templateDir, flagSpec string) (*model.Spec, error) {
+	buf, err := fs.ReadFile(filepath.Join(templateDir, flagSpec))
+	if err != nil {
+		return nil, fmt.Errorf("error reading template spec: ReadFile(): %w", err)
+	}
+
+	decoder := model.NewDecoder(bytes.NewReader(buf))
```

Comment:
```go
f, err := fs.Open(filepath.Join(templateDir, flagSpec))
if err != nil {
  // ...
}
defer f.Close()

decoder := model.NewDecoder(f)
```

? Avoids reading the entire thing into a buffer.

Reply:
Yeah, that is better. Done.